### PR TITLE
Update TraceLogging to use NULL-terminated strings instead of counted strings

### PR DESCRIPTION
**Pull request by brianrob**
Contributes to #11939.

Within EventSource, TraceLogging events write counted strings while manifest-based events write NULL-terminated strings.  This change converts TraceLogging to write NULL-terminated strings so that only one format needs to be supported when writing events to EventPipe.

### CODE REVIEW

**Comment by brianrob** [comment]
Run EventSource tests: @dotnet-bot test Windows_NT x64 Checked corefx_baseline

File: src/mscorlib/shared/System/Diagnostics/Tracing/TraceLogging/DataCollector.cs [hunk 1]
```diff
@@ -143,6 +143,51 @@ internal void AddBinary(string value, int size)
...
+            {
+                if (this.bufferNesting == 0)
+                {
+                    this.ScalarsEnd();
+                    this.PinArray(value, size);
+                }
+                else
+                {
+                    var oldPos = this.bufferPos;
+                    this.bufferPos = checked(this.bufferPos + size);
```
Comment (vancem) [inline src/mscorlib/shared/System/Diagnostics/Tracing/TraceLogging/DataCollector.cs]: Make sure we have some tests that have strings in nested structures (we probably do) this insures that we tests both of these code paths.

File: src/mscorlib/shared/System/Diagnostics/Tracing/TraceLogging/TraceLoggingMetadataCollector.cs [hunk 2]
```diff
@@ -192,22 +192,44 @@ public void AddBinary(string name, TraceLoggingDataType type)
...
             {
                 case TraceLoggingDataType.Utf16String:
                 case TraceLoggingDataType.MbcsString:
+                    break;
+                default:
+                    throw new ArgumentOutOfRangeException(nameof(type));
+            }
+
+            this.impl.AddNonscalar();
+            this.AddField(new FieldMetadata(name, type, this.Tags, this.BeginningBufferedArray));
```
Comment (vancem) [inline src/mscorlib/shared/System/Diagnostics/Tracing/TraceLogging/TraceLoggingMetadataCollector.cs]: I can't say that this is correct or not.  Ideally Doug validates this.

Comment (brianrob) [reply src/mscorlib/shared/System/Diagnostics/Tracing/TraceLogging/TraceLoggingMetadataCollector.cs]: Thanks @vancem.  I am working with Doug offline to review this change.  I also am investigating the CI failure which looks like a real issue.

File: src/mscorlib/shared/System/Diagnostics/Tracing/TraceLogging/DataCollector.cs (hunk 1 above)
Comment (brianrob) [reply src/mscorlib/shared/System/Diagnostics/Tracing/TraceLogging/DataCollector.cs]: I have confirmed that we do have tests for this and that they are passing.

**Comment by brianrob** [comment]
The latest changes include feedback from Doug as well as a fix for the functional test break.

**Comment by brianrob** [comment]
Run EventSource tests: @dotnet-bot test Windows_NT x64 Checked corefx_baseline

**Comment by brianrob** [comment]
Infrastructure failure: @dotnet-bot test OSX10.12 x64 Checked Innerloop Build and Test

**Comment by brianrob** [comment]
Infrastructure failure: @dotnet-bot test OSX10.12 x64 Checked Innerloop Build and Test

**Comment by brianrob** [comment]
Infrastructure failure: @dotnet-bot test OSX10.12 x64 Checked Innerloop Build and Test

**Comment by brianrob** [comment]
@dotnet-bot test Windows_NT x64 Checked Build and Test (Jit - CoreFx)

**Comment by brianrob** [comment]
Run EventSource tests: @dotnet-bot test Windows_NT x64 Checked corefx_baseline

**Comment by brianrob** [comment]
OSX failures due to https://github.com/dotnet/core-eng/issues/2808.
CoreFx failures due to APIs that are out of sync between CoreCLR and CoreFx.

**Comment by jkotas** [comment]
You have set executable bit on some of the files. It is the `100644 → 100755` in the github diff. I do not think you meant to do that.

**Pull request merged**